packages/client/src/pages/Game/engine/TDEngine.tsx

```diff
@@ -0,0 +1,770 @@
+        this.towerTwoParam.projectileParams,
+      );
+    }
+  };
+
+  public buildThirdTower = () => {
+    if (this.isEnoughMoney(this.towerThreeParam.towerParams.price)) {
+      this.isCanBuild = true;
+      this.draftTower = new Tower(
+        this,
+        this.towerSprites.levelThree!,
+        this.projectileSprites.levelThree!,
+        this.projectileHitSprites.levelThree!,
+        this.draftBuildCoordinates,
+        this.towerThreeParam.towerParams,
+        this.towerThreeParam.projectileParams,
+      );
+    }
+  };
+
+  public findClosestTile(coordinates: ITwoDCoordinates) {
+    let minDistance = this.map?.mapParams.width;
+    for (let tile of this.map?.mapParams.mapTilesArr!) {
+      const distance =
+        (tile.x -
+          coordinates.x! +
+          this.map?.mapParams.gridStep! -
+          this.map?.mapParams?.tileCenter!) *
+          (tile.x -
+            coordinates.x! +
+            this.map?.mapParams.gridStep! -
+            this.map?.mapParams?.tileCenter!) +
+        (tile.y -
+          coordinates.y! +
+          this.map?.mapParams.gridStep! +
+          this.map?.mapParams?.tileCenter!) *
+          (tile.y -
+            coordinates.y! +
+            this.map?.mapParams.gridStep! +
+            this.map?.mapParams?.tileCenter!);
+      if (distance < minDistance!) {
+        minDistance = distance;
+        this.map!.mapParams.closestTile! = tile!;
+      }
+    }
+
+    this.draftBuildCoordinates = {
+      x: this.map?.mapParams.closestTile.x! + this.map?.mapParams.gridStep!,
+      y: this.map?.mapParams.closestTile.y! + this.map?.mapParams.gridStep!,
+    };
+
+    return {
+      x: this.map?.mapParams.closestTile.x,
+      y: this.map?.mapParams.closestTile.y!,
+    };
+  }
+
+  highlightTile(coords: ITwoDCoordinates) {
+    this.context?.beginPath();
+    this.context!.strokeStyle = "green";
+    this.context?.setLineDash([]);
+    this.context?.strokeRect(
+      coords.x,
+      coords.y,
+      this.map?.mapParams.gridStep!,
+      this.map?.mapParams.gridStep!,
+    );
+    this.context?.closePath();
+  }
+
+  public canvasMouseMoveCallback = (e: MouseEvent) => {
+    this.cursorPosition = { x: e.pageX, y: e.pageY };
+    this.findClosestTile(this.cursorPosition);
+    this.draftShowTower();
+  };
+
+  public canvasClickCallback = (e: MouseEvent) => {
+    this.draftBuildTower();
+  };
+
+  public gameWindowKeydown = (e: KeyboardEvent) => {
+    this.manageHotkeys(e);
+  };
+
+  public isEnoughMoney(towerPrice: ITower["towerParams"]["price"]) {
+    if (this.money >= towerPrice!) {
+      return true;
+    }
+    this.isNotEnoughMoney = true;
+    return false;
+  }
+
+  public draftShowTower() {
+    if (this.isCanBuild) {
+      // show building grid
+      // this.isShowGrid = true
+
+      if (!this.draftTower) {
+        this.draftTower = new Tower(
+          this,
+          undefined,
+          undefined,
+          undefined,
+          this.draftBuildCoordinates,
+        );
```

Comment:
if you could pass some params as undefined, wouldn't it be easier to pass params as a single options object?
```suggestion
        const engine = this;
        this.draftTower = new Tower({
          engine,
          draftBuildCoordinates: engine.draftBuildCoordinates,
        });
```

Reply:
Skipped, theese params are required, but as Tower draft you can pass just undefined, they are being populated in the future